elf-size-analyze.py

```diff
@@ -895,6 +900,31 @@ def _generate_protolines(self, min_size):
             protolines.append(self.Protoline(depth, node))
         return protolines
 
+    def _generate_node_dict(self, min_size):
+        # generate dict of nodes
+        nodeDict = dict()
+        for node, depth in self.tree_root.pre_order():
+            nd = NestedDict()
+            if node.is_root():
+                continue
+            elif not (node.is_symbol() or node.is_path()):
+                raise Exception('Wrong symbol type encountered')
+            elif node.is_symbol() and node.data.size < min_size:
+                continue
+            nodePath = list()
+            iterNode = node
+            nodePath.insert(0, iterNode.data.name if iterNode.is_symbol() else iterNode.data)
+            while iterNode.parent.data != None:
+                nodePath.insert(0, "children")
+                nodePath.insert(0, iterNode.parent.data.name if iterNode.parent.is_symbol() else iterNode.parent.data)
+                iterNode = iterNode.parent    
+
+            nd[tuple(nodePath)+("name",)] = node.data.name if node.is_symbol() else node.data
+            nd[tuple(nodePath)+("cumulative_size",)] = node.cumulative_size
+            merge(nodeDict, nd.to_dict())            
```

Comment:
Because we are iterating over the tree in pre-order, it's possible to avoid merging dictionaries here (because pre-order iterates from lowest depth to highest). This way we could avoid dependency on `ndicts` and `mergedeep` with minimal code changes. The following should produce identical result:
```python
def _generate_node_dict(self, min_size):
    # generate dict of nodes
    nodeDict = dict()
    get_key = lambda node: node.data.name if node.is_symbol() else node.data

    for node, depth in self.tree_root.pre_order():
        if node.is_root():
            continue
        elif not (node.is_symbol() or node.is_path()):
            raise Exception('Wrong symbol type encountered')
        elif node.is_symbol() and node.data.size < min_size:
            continue

        nodePath = list()
        iterNode = node
        while iterNode.parent is not None:
            nodePath.append(iterNode)
            iterNode = iterNode.parent
        nodePath.reverse()

        children = nodeDict
        for n in nodePath[:-1]:
            children = children[get_key(n)]['children']

        key = get_key(node)
        children[key] = {
            'name': get_key(node),
            'cumulative_size': node.cumulative_size,
        }
        if not node.is_symbol():
            children[key]['children'] = {}

    return nodeDict
```